#422 Added support for building rpm and AppImage installers. Moved generating deb and tar.gz from fpm to eletron-builder

natanielcz opened the pull request.

dmsnell commented:
This looks like some great stuff @natanielcz - can you help us by providing some background on the changes? It looks like there's far more in play than just adding RPM support.

natanielcz commented:
I've updated PR title. Generally, I added support for two new formats: rpm and AppImage. And move deb and tar.gz packaging from fpm to electron-builder. To make it easier for you during releasing a new version I updated makefile and added building deb, rpm, tar.gz, and AppImage to target `linux`.

This pull request refers also to #530.

I think that we should also consider moving Windows and Mac installers to electron-builder to have it all in one place. 

However, there is still a bug in electron-builder electron-userland/electron-builder#1907 and application icon isn't visible in the Software / Ubuntu Store. Electron builder will allow us to make also Snap installer for Linux, DMG, MAS (Mac Application Store) and PKG for Mac, MSI installer, NSIS Installer and Web Installer but also AppX (for Windows Store) for Windows.

natanielcz commented:
One more fix. We don't need copy icons manually so I've deleted that for Linux build.

rodrigoi commented on electron-builder.json:
```diff
@@ -0,0 +1,37 @@
+{
+  productName: "Simplenote",
```
would you mind adding quotes to the keys on this file?

rodrigoi commented:
@natanielcz would you mind rebasing this PR? Thanks!

natanielcz commented:
I'll take care of that after 8th of December.

natanielcz commented:
@rodrigoi: Rebased.

rodrigoi commented:
thanks @natanielcz !! I have a couple comments:

I don't have `electron-builder` as a global. To `make package-linux` work I've renamed `-PACKAGE_WIN32` to `ELECTRON_BUILDER` and used it instead of `@electron-builder`. It was also complaining about `projectDir`, so the command looked like:

```make
package-linux: linux
	@$(ELECTRON_BUILDER) --linux --x64
```

We usually build on macOS, and `rpm` fails with a warning of a missing dependency. Maybe we should add `brew install rpm` to the documentation.

And finally, `yarn.lock` is creating an artificially large diff, and at this moment, this project does not use `yarn`. You may want to create a new PR to add `yarn`.

Besides that, this tests very well. All packages work (except for the image on Software/Ubuntu Store you've [mentioned before](https://github.com/Automattic/simplenote-electron/pull/650#issuecomment-341082624))

I'll try to get another set of 👀 to review this, but this looks really promising. Thanks for the hard work!

natanielcz commented:
@rodrigoi You're right that electron-builder installed globally is currently required but I'm working on it. I removed electron-builder from global dependencies and will test a little bit different approach.

natanielcz commented:
@rodrigo, @roundhill @dmsnell : I've also created `.snap` the installer which could be put in the [snapcraft store](www.snapcraft.io). Give me your emails and I'll add you to collaborators to my test Simplenote app on snapcraft so You can see and test installation from there.

roundhill commented:
Seems to be working quite well 👍 I did get this warning when attempting to install the .deb:

<img width="595" alt="screen shot 2018-03-06 at 4 04 48 pm" src="https://user-images.githubusercontent.com/789137/37066051-75964f96-2158-11e8-86bd-59fe503209b0.png">

natanielcz commented:
Lintian returns more thanks that one error. I'll compare my deb with deb installed from the website to check what else should be done

natanielcz commented:
@roundhill which app did you use to install this deb? I can't reproduce that issue.

roundhill commented:
> @roundhill which app did you use to install this deb? I can't reproduce that issue.

I used whatever the default Ubuntu installer is, is that Ubuntu Software Center? Sorry, I'm not much of a linux person :)

roundhill commented:
My VM is actually on an older version of Ubuntu (15), maybe it's not an issue in newer releases?

natanielcz commented:
@roundhill You may be right about that. I have Ubuntu 17.10 and I don't see anything like that but I'll try to fix at least critical linter problems. It may take a while due to some electron limitations and my lack of knowledge about loading system libraries.

roundhill commented:
OK, I will update it and let you know if I see the same error.

roundhill commented:
> OK, I will update it and let you know if I see the same error.

No error on 17.10 👍 

Can you remove the `projectDir` argument that @rodrigoi mentioned? It doesn't appear to be needed, unless I'm missing something. It prevents me from building on macOS without removing it.

natanielcz commented:
I removed that already

roundhill commented:
I think this is good to go. I've tested the builds and they work great.
